### ISSUE TITLE
Error When Migrating Policies with LDAP Group Limitations

### ISSUE THREAD

**Issue by sjgrall** [jamf/JamfMigrator#53]
When attempting to migrate policies with LDAP group(s) set on the Limitations tab of the Jamf Pro policy, the following type of error is returned, and the policy is not migrated:

20200513_075250 [CreateEndpoints] [policies] Security - Reissue FileVault Recovery Key - Failed (409).  Create error: Problem matching limit_to_users user group .

All other policies (policies without LDAP group limitations) were migrated. This error appears to have been returned for every single policy whereby we had set an LDAP group limitation.

**Comment by BIG-RAT** [comment]
Thanks for the heads up @sjgrall.  Issue should be resolved in v5.2.5

**Comment by sjgrall** [comment]
It's still not working, unfortunately. I get a different response now:
20200513_192228 [CreateEndpoints] [policies] Security - Reissue FileVault Recovery Key - Failed (409).  Create error: Problem matching limitation user group .

**Comment by BIG-RAT** [comment]
I trust the LDAP group(s) appear when tested in the LDAP config?  Could you post what you have (raw XML) for the scope of one of the failing policies?  My testing with the following had no issues.
```
<scope>
	<all_computers>false</all_computers>
	<computers>
		<computer>
			<id>54</id>
			<name>dev-rmbp</name>
			<udid>689349F2-3D36-1B66-3EFA-BBAED37066F7</udid>
		</computer>
	</computers>
	<computer_groups>
		<computer_group>
			<id>182</id>
			<name>East</name>
		</computer_group>
	</computer_groups>
	<buildings>
		<building>
			<id>1</id>
			<name>Corporate</name>
		</building>
	</buildings>
	<departments></departments>
	<limit_to_users>
		<user_groups>
			<user_group>PowerUsers</user_group>
		</user_groups>
	</limit_to_users>
	<limitations>
		<users></users>
		<user_groups>
			<user_group>
				<id>16536</id>
				<name>PowerUsers</name>
			</user_group>
		</user_groups>
		<network_segments></network_segments>
		<ibeacons></ibeacons>
	</limitations>
	<exclusions>
		<computers></computers>
		<computer_groups></computer_groups>
		<buildings></buildings>
		<departments></departments>
		<users>
			<user>
				<name>Joe90</name>
			</user>
		</users>
		<user_groups>
			<user_group>
				<id>16520</id>
				<name>Remote</name>
			</user_group>
		</user_groups>
		<network_segments></network_segments>
		<ibeacons></ibeacons>
	</exclusions>
</scope>
```

**Comment by sjgrall** [comment]
Yes, LDAP groups appear if tested. How do I generate that XML output?

**Comment by BIG-RAT** [comment]
Click the preferences button, export tab, then select the desired export options.
Raw is the source XML as it is pulled from the server.
Trimmed is the XML that is posted to the destination server.
And of course be sure to include the scope.

<img width="405" alt="export" src="https://user-images.githubusercontent.com/22251543/81978309-54c7d880-95f9-11ea-8804-f9e1d620cc89.png">

**Comment by sjgrall** [comment]
Here's a RAW XML example, as you requested:

<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<policy>
    <general>
        <name>App - MacVector - v 17.5</name>
        <enabled>true</enabled>
        <trigger>USER_INITIATED</trigger>
        <trigger_checkin>false</trigger_checkin>
        <trigger_enrollment_complete>false</trigger_enrollment_complete>
        <trigger_login>false</trigger_login>
        <trigger_logout>false</trigger_logout>
        <trigger_network_state_changed>false</trigger_network_state_changed>
        <trigger_startup>false</trigger_startup>
        <trigger_other></trigger_other>
        <frequency>Ongoing</frequency>
        <location_user_only>false</location_user_only>
        <target_drive>/</target_drive>
        <offline>false</offline>
        <category>
            <name>Applications</name>
        </category>
        <date_time_limitations>
            <activation_date></activation_date>
            <activation_date_epoch>0</activation_date_epoch>
            <activation_date_utc></activation_date_utc>
            <expiration_date></expiration_date>
            <expiration_date_epoch>0</expiration_date_epoch>
            <expiration_date_utc></expiration_date_utc>
            <no_execute_on></no_execute_on>
            <no_execute_start></no_execute_start>
            <no_execute_end></no_execute_end>
        </date_time_limitations>
        <network_limitations>
            <minimum_network_connection>No Minimum</minimum_network_connection>
            <any_ip_address>true</any_ip_address>
            <network_segments></network_segments>
        </network_limitations>
        <override_default_settings>
            <target_drive>default</target_drive>
            <distribution_point></distribution_point>
            <force_afp_smb>false</force_afp_smb>
            <sus>default</sus>
            <netboot_server>current</netboot_server>
        </override_default_settings>
        <network_requirements>Any</network_requirements>
        <site>
            <name>None</name>
        </site>
    </general>
    <scope>
        <all_computers>true</all_computers>
        <computers></computers>
        <computer_groups></computer_groups>
        <buildings></buildings>
        <departments></departments>
        <limitations>
            <users></users>
            <user_groups>
                <user_group>
                    <name>NCI CBIIT Desktop Platform Management</name>
                </user_group>
                <user_group>
                    <name>NCI CBIIT Desktop Technicians</name>
                </user_group>
                <user_group>
                    <name>NCI DEA Secondary Administrators</name>
                </user_group>
            </user_groups>
            <network_segments></network_segments>
            <ibeacons></ibeacons>
        </limitations>
        <exclusions>
            <computers></computers>
            <computer_groups></computer_groups>
            <buildings></buildings>
            <departments></departments>
            <users></users>
            <user_groups></user_groups>
            <network_segments></network_segments>
            <ibeacons></ibeacons>
        </exclusions>
    </scope>
    <self_service>
        <use_for_self_service>true</use_for_self_service>
        <self_service_display_name>MacVector 17.5</self_service_display_name>
        <install_button_text>Install</install_button_text>
        <reinstall_button_text>Reinstall</reinstall_button_text>
        <self_service_description>This policy will install MacVector 17.5 on this workstation.</self_service_description>
        <force_users_to_view_description>true</force_users_to_view_description>
        <feature_on_main_page>false</feature_on_main_page>
        <self_service_categories>
            <category>
                <name>Applications</name>
                <display_in>true</display_in>
                <feature_in>false</feature_in>
            </category>
        </self_service_categories>
        <notification>false</notification>
        <notification>Self Service</notification>
        <notification_subject>MacVector 17.01</notification_subject>
        <notification_message></notification_message>
    </self_service>
    <package_configuration>
        <packages>
            <size>1</size>
            <package>
                <name>App - MacVector 17.5.pkg</name>
                <action>Install</action>
                <fut>false</fut>
                <feu>false</feu>
                <update_autorun>false</update_autorun>
            </package>
        </packages>
    </package_configuration>
    <scripts>
        <size>0</size>
    </scripts>
    <printers>
        <size>0</size>
        <leave_existing_default></leave_existing_default>
    </printers>
    <dock_items>
        <size>0</size>
    </dock_items>
    <account_maintenance>
        <accounts>
            <size>0</size>
        </accounts>
        <directory_bindings>
            <size>0</size>
        </directory_bindings>
        <management_account>
            <action>doNotChange</action>
        </management_account>
        <open_firmware_efi_password>
            <of_mode>none</of_mode>
            <of_password_sha256 since="9.23">e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855</of_password_sha256>
        </open_firmware_efi_password>
    </account_maintenance>
    <reboot>
        <message>This computer will restart in 5 minutes. Please save anything you are working on and log out by choosing Log Out from the bottom of the Apple menu.</message>
        <startup_disk>Current Startup Disk</startup_disk>
        <specify_startup></specify_startup>
        <no_user_logged_in>Restart if a package or update requires it</no_user_logged_in>
        <user_logged_in>Restart if a package or update requires it</user_logged_in>
        <minutes_until_reboot>5</minutes_until_reboot>
        <start_reboot_timer_immediately>false</start_reboot_timer_immediately>
        <file_vault_2_reboot>false</file_vault_2_reboot>
    </reboot>
    <maintenance>
        <recon>false</recon>
        <reset_name>false</reset_name>
        <install_all_cached_packages>false</install_all_cached_packages>
        <heal>false</heal>
        <prebindings>false</prebindings>
        <permissions>false</permissions>
        <byhost>false</byhost>
        <system_cache>false</system_cache>
        <user_cache>false</user_cache>
        <verify>false</verify>
    </maintenance>
    <files_processes>
        <search_by_path></search_by_path>
        <delete_file>false</delete_file>
        <locate_file></locate_file>
        <update_locate_database>false</update_locate_database>
        <spotlight_search></spotlight_search>
        <search_for_process></search_for_process>
        <kill_process>false</kill_process>
        <run_command></run_command>
    </files_processes>
    <user_interaction>
        <message_start></message_start>
        <allow_users_to_defer>false</allow_users_to_defer>
        <allow_deferral_until_utc></allow_deferral_until_utc>
        <message_finish></message_finish>
    </user_interaction>
    <disk_encryption>
        <action>none</action>
    </disk_encryption>
</policy>

**Comment by BIG-RAT** [comment]
Thanks @sjgrall - I created the same groups (in case there was something about the name) and scope.  Migration worked worked for me.
Here is where we might need the waders:
If you could, put your Jamf server into debug mode and then try to migrate again.  In the debug log search for LDAP or part of a group name.  Should see something like the following:
```
2020-05-15 09:55:45,138 [DEBUG] [Thread-176 ] [DefaultLDAPLookupService ] - Search LDAPServer [ID=4, Name=lab.private] for group NCI CBIIT Desktop Technicians with wildcards true
2020-05-15 09:55:45,139 [DEBUG] [Thread-176 ] [DefaultLDAPLookupService ] - Search Filter: (&(&(objectClass=top)(objectClass=group))(name=*NCI CBIIT Desktop Technicians*))
2020-05-15 09:55:45,139 [DEBUG] [Thread-176 ] [DefaultLDAPLookupService ] - Open LDAP Connection to lab.private
2020-05-15 09:55:45,161 [DEBUG] [Thread-176 ] [DefaultLDAPLookupService ] - Executing LDAP search on base 'DC=lab,DC=private' with scope 2
2020-05-15 09:55:45,170 [DEBUG] [Thread-176 ] [DefaultLDAPLookupService ] - CN=NCI CBIIT Desktop Technicians,OU=Remote User Test,OU=Another OU,OU=Test,OU=Corp: null:null:{name=name: NCI CBIIT Desktop Technicians, instancetype=instanceType: 4, grouptype=groupType: -2147483644, objectsid=objectSid: [B@779b8335, samaccounttype=sAMAccountType: 536870912, usncreated=uSNCreated: 2796634, usnchanged=uSNChanged: 2796634, objectclass=objectClass: top, group, distinguishedname=distinguishedName: CN=NCI CBIIT Desktop Technicians,OU=Remote User Test,OU=Another OU,OU=Test,OU=Corp,DC=lab,DC=private, objectcategory=objectCategory: CN=Group,CN=Schema,CN=Configuration,DC=lab,DC=private, samaccountname=sAMAccountName: NCI CBIIT Desktop Technicians, objectguid=objectGUID: [B@77994e78, cn=cn: NCI CBIIT Desktop Technicians, whencreated=whenCreated: 20200515143409.0Z, whenchanged=whenChanged: 20200515143409.0Z, dscorepropagationdata=dSCorePropagationData: 16010101000000.0Z}
2020-05-15 09:55:45,170 [DEBUG] [Thread-176 ] [DefaultLDAPLookupService ] - LDAP parse Group
2020-05-15 09:55:45,171 [DEBUG] [Thread-176 ] [DefaultLDAPLookupService ] - LDAP parse Group done
2020-05-15 09:55:45,171 [DEBUG] [Thread-176 ] [DefaultLDAPLookupService ] - PartialResultException was handled.
2020-05-15 09:55:45,171 [DEBUG] [Thread-176 ] [DefaultLDAPLookupService ] - Finished LDAP lookup
2020-05-15 09:55:45,172 [DEBUG] [Thread-176 ] [DefaultLDAPLookupService ] - Closing LDAP Connect
2020-05-15 09:55:45,172 [DEBUG] [Thread-176 ] [DefaultLDAPLookupService ] - Search LDAPServer [ID=4, Name=lab.private] for group NCI DEA Secondary Administrators with wildcards true
2020-05-15 09:55:45,172 [DEBUG] [Thread-176 ] [DefaultLDAPLookupService ] - Search Filter: (&(&(objectClass=top)(objectClass=group))(name=*NCI DEA Secondary Administrators*))
2020-05-15 09:55:45,172 [DEBUG] [Thread-176 ] [DefaultLDAPLookupService ] - Open LDAP Connection to lab.private
2020-05-15 09:55:45,173 [DEBUG] [Thread-176 ] [DefaultLDAPLookupService ] - Executing LDAP search on base 'DC=lab,DC=private' with scope 2
2020-05-15 09:55:45,177 [DEBUG] [Thread-176 ] [DefaultLDAPLookupService ] - PartialResultException was handled.
2020-05-15 09:55:45,177 [DEBUG] [Thread-176 ] [DefaultLDAPLookupService ] - Finished LDAP lookup
2020-05-15 09:55:45,178 [DEBUG] [Thread-176 ] [DefaultLDAPLookupService ] - Closing LDAP Connect
```
Here, you see it found the group NCI CBIIT Desktop Technicians (timestamp: 2020-05-15 09:55:45,170) but did not find the group NCI DEA Secondary Administrators (I'd deleted this group) and gave up on any further lookups.
I'm curious if you see any successful lookups.  Ideally all should return something like
```
2020-05-15 09:55:45,170 [DEBUG] [Thread-176 ] [DefaultLDAPLookupService ] - CN=NCI CBIIT Desktop Technicians,OU=Remote User Test,OU=Another OU,OU=Test,OU=Corp: null:null:{name=name: NCI CBIIT Desktop Technicians, instancetype=instanceType: 4, grouptype=groupType: -2147483644, objectsid=objectSid: [B@779b8335, samaccounttype=sAMAccountType: 536870912, usncreated=uSNCreated: 2796634, usnchanged=uSNChanged: 2796634, objectclass=objectClass: top, group, ...
```

**Comment by sjgrall** [comment]
Yes, the lookups are successful on the destination server, and it returns all group members for each of those groups. However, because of the sensitive nature of the group membership, I cannot send you the exact output from the debug log.

**Comment by sjgrall** [comment]
Update: The remaining issue I've had was caused by the presence of two LDAP server configs with overlapping matching user groups. Our institution's AD is unique in its structure, and some user groups are accessible from two different OU structures on the same AD LDAP system, whereas others are not. I removed the secondary LDAP server config, which isn't needed on our dev server, and it worked around the issue.
